workflows/bot: migrate nixpkgs-merge-bot to GHA

wolfgangwalther opened the pull request:
Running the nixpkgs-merge-bot in GitHub Actions instead of a separate workflow has multiple advantages:
- A much better development workflow, with improved testability.
- The ability to label PRs with a "merge-bot eligible" label from the same codebase.
- Using more data for merge strategy decisions, for example the number of rebuilds.

This commits re-implements most of the features from the current nxipkgs-merge-bot directly in the bot workflow. Instead of reacting to webhook events, this now runs on the regular 10 minute schedule. Some merges might be delayed a few minutes, but that should not be a problem in practice.

To give the user early feedback, there are additional workflows running when a comment or review is posted. These react with "eyes" to make the user aware that the comment has been recognized.

The only feature not taken over was the size check for files in the PR. This kind of check is not really relevant for maintainer merges only - if we want to prevent bigger files from making it into the tree, then we need a generic CI check, which is out of scope for the merge-bot.

Other than that, everything should be implemented - any omissions are by accident.

Resolves https://github.com/NixOS/nixpkgs-merge-bot/issues/211, but not https://github.com/NixOS/nixpkgs/issues/378214, yet.

Also resolves https://github.com/NixOS/nixpkgs-merge-bot/issues/231, because we only use two merge methods, of which auto-merge needs to be tried optimistically.

Also resolves https://github.com/NixOS/nixpkgs-merge-bot/issues/148, because the scheduled approach retries automatically.

## Things done

- [x] Tested eyes reactions in my fork.
- [x] Tested successful and failing merges locally.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
workflows/bot: rename from labels

This workflow / script is already doing more than must labeling: it's
already auto-closing package request issues.

Since we're going to migrate the nixpkgs-merge-bot into this workflow,
we'll rename things to a more generic name.

wolfgangwalther committed:
workflows/bot: migrate nixpkgs-merge-bot to GHA

Running the nixpkgs-merge-bot in GitHub Actions instead of a separate
workflow has multiple advantages:
- A much better development workflow, with improved testability.
- The ability to label PRs with a "merge-bot eligible" label from the
same codebase.
- Using more data for merge strategy decisions, for example the number
of rebuilds.

This commits re-implements most of the features from the current
nxipkgs-merge-bot directly in the bot workflow. Instead of reacting to
webhook events, this now runs on the regular 10 minute schedule. Some
merges might be delayed a few minutes, but that should not be a problem
in practice.

To give the user early feedback, there are additional workflows running
when a comment or review is posted. These react with "eyes" to make the
user aware that the comment has been recognized.

The only feature not taken over was the size check for files in the PR.
This kind of check is not really relevant for maintainer merges only -
if we want to prevent bigger files from making it into the tree, then we
need a generic CI check, which is out of scope for the merge-bot.

Other than that, everything should be implemented - any omissions are by
accident.

emilazy reviewed (commented):
Thank you for working on this! Very much in support of glue like this living inside the monorepo where it’s accessible to contribute to and can reuse our common CI machinery, and I’m excited to see it extended for [trusted maintainers](https://discourse.nixos.org/t/nixpkgs-core-team-update-2025-10-10/70730) once the details of that proposal are fleshed out and go through community review :)

I don’t know the implementation details of the current merge bot, so feel free to dismiss any of these comments that are unchanged from that.

wolfgangwalther committed:
workflows/{review,reviewed}: rename from dismissed-review and review-dismissed

This is just a separate commit for a better diff. This workflow will be
extended with the `submitted` type in the next commit, so needs a
different naming.

While at it, also remove the useless concurrency group: This workflow is
never run in a PR context, so will never be cancelled.

wolfgangwalther committed:
workflows/bot: retry failed merges

By not keeping the node_id in the comments resulting from a failed
merge, these merges will be automatically retried.

MattSturgeon commented on .github/workflows/bot.yml:
```diff
@@ -3,7 +3,7 @@
 # a way that allows code injection.
 
-name: Labels
+name: Bot
```
Kinda awkward, but this thread is now discussing a) the name for the workflow, b) the gh-app to use to merge PRs, and c) whether we can edit the merge-commit message.

Re a): I'm not really keen on any of the names suggested so far (bot, lifecycle, sync), but I think I like "bot" a tiny bit less than the others.

Re b): my preference would be to continue using the merge-bot app to merge PRs. I understand why this causes issues with (things like) ratelimits, though.

Don't we already deal with having multiple tokens (ci, owner, gh-actions) in some other workflows, though?

If we're only using the merge-bot for the actual enqueue mutation, then I doubt we'll need to worry too much about its effect on ratelimits for now.

Re c): see https://github.com/NixOS/nixpkgs-merge-bot/issues/100#issuecomment-3302532147

wolfgangwalther replied:
> I don’t associate the nixpkgs-ci account adjusting labels with the word “bot” at all. I only thought of the merge bot when seeing this – and of course those are two separate names, actually. Labels being adjusted by a job called “Bot” isn’t intuitive to me.

Odd. This is a very intuitive association for me. ofborg was handling the labels earlier, ofborg is a bot. We have implemented our own bot, GHA based. GitHub shows these apps as `nixpkgs-ci[bot]` - it even renders that little `bot` tag.

<img width="578" height="46" alt="2025-10-30T14:37:05,918547995+01:00" src="https://github.com/user-attachments/assets/c5918711-0a2e-4529-9b97-ef23f81b46f2" />
<img width="233" height="39" alt="2025-10-30T14:37:15,784117671+01:00" src="https://github.com/user-attachments/assets/9d32f904-7f1e-42f5-b66a-7c675423ba7c" />

I think "bot" is actually the *correct* term from a GitHub perspective here.

Yes, we have *some* bot activity in other workflows, such as periodic merges, reviewer requests (as of now, I eventually want to move them to the "bot" workflow as well), backports, team sync. But realistically, after doing the merge-bot part, *most* of what the bot does is in that one workflow.

I'd be happy to rename the other workflows to add a `bot-` prefix, so that it's clear we have all the "automation" stuff named that way.

wolfgangwalther commented:
> After talking through various failure modes with @emilazy, I marked this as draft. Will try to implement something not based on the scheduled workflow, but run directly in the PR.

This did not turn out well, because I'm hitting many more complications when trying to run this in each PR.

Here's a summary of what happened since. I discussed the following failure modes / improvements with Emily:
- The merge bot would provide better UX if the job that actually merges after the comment is posted, would run in the PR's context - and thus appear in the PR's checklist. If it fails - the user would have a visible link to the job log immediately, in the PR's checklist. That's a nice idea indeed, but hard to put in practice: To get the actual job to run in the PRs context, the only way I found was to use a job that is triggered on the regular `pull_request_target` events - and then to *re-run* that job on merge. Initially it would just skip execution, but when re-run it would pick up the merge comment and act. While this would all work - it would also mean that the related workflow file would always stay the one that was currently on the target branch *when that PR event happened*. So for a PR that was opened a few months ago, this would always use the merge bot logic from a few months ago - and fixing merge bot logic afterwards would require re-running CI. In some cases this might be annoying - in other cases (when the merge bot merges more than it should) it's a security risk. In the end I decided against that approach - we can still look into posting status checks in the check list instead of / in addition to the comments we currently do, later.
- The current implementation in this PR always takes the maintainers who are allowed to merge from the PR's Eval results. That means they might be quite outdated for older PRs. This is a regression compared to the current merge-bot, which always takes the latest maintainer info from the target branch (or always master? not sure). It is also a potential security problem, because in cases of abuse, there is no quick way of retracting someone's merge permissions via the bot, without re-running CI for all related PRs. We discussed the following to deal with that:
  - First of all, the (new) merge-bot should check whether the maintainer is actually part of the nixpkgs-maintainers team, before merging. This makes sense, because all our privileges ("write" for committers) are actually tied to GitHub - so we should do the same check for maintainers. This will allow to remove a maintainer from that team and thus revoking their privileges immediately - without having to propagate the change to the maintainers list to all target branches or even to all PRs. This is a limitation of the current merge-bot, too, and is independent of whether we run this in the PR or via schedule in the future.
  - Furthermore, we should always take the maintainer information from the target branch *at the time of posting the comment*, not from when Eval on the PR ran. Doing this in a PR-based workflow, would have made this easy, because that could have just checked out the target branch and ran Nix on it, similar to the current nixpkgs-merge-bot implementation. In the scheduled workflow, this is not as easy, because this job would potentially have to checkout a few target branches, as it works on multiple PRs at once. This, plus the installation of Nix - will take time, which we should really try to limit for a job that runs every 10 minutes. Thus, I came up with #456965, which will just store a map of maintainers as artifact on the target branch latest workflow run. The scheduled job can download this and use up-to-date maintainer information - not just for the merge-bot, but later also for reviewer requests and labeling itself.

With this in mind, I will proceed with the approach in this PR in general. I will most likely keep the scheduled approach compared to running the merge-bot in the `issue_comment` or `workflow_run` (via `pull_request_review`) events directly. While the latter would provide more instant UX, I'll have to see whether it would actually allow to simplify things.. or whether they'd actually become more complex instead.

wolfgangwalther commented:
> * First of all, the (new) merge-bot should check whether the maintainer is actually part of the nixpkgs-maintainers team, before merging.
> [...]
> * Furthermore, we should always take the maintainer information from the target branch _at the time of posting the comment_, not from when Eval on the PR ran.

Both of these are now implemented. Feature-wise this is ready, I think.

wolfgangwalther commented:
The new bot had its first successful merge in https://github.com/NixOS/nixpkgs/pull/457465.

The bot will also react to PRs that the old merge bot had previously been asked to merge, but failed. Some examples are:
- https://github.com/NixOS/nixpkgs/pull/455776#issuecomment-3476474702
- https://github.com/NixOS/nixpkgs/pull/457336#issuecomment-3476474662

This will happen once, because the new bot doesn't recognize the old bot's comments. Not a problem.
